### PR TITLE
[URP]Quality audit-Material editor help button

### PR DESCRIPTION
**Pull request by alex-vazquez**
---
### Purpose of this PR
https://jira.unity3d.com/browse/XPIPELINE-191

All help buttons point to-> https://docs.unity3d.com/Packages/com.unity.render-pipelines.universal@11.0/manual/shaders-in-universalrp.html

image

---
### Testing status
Check that the button opens the correct url

---
### Comments to reviewers
@TheoWong-pixel The solution where each shader changes it's url is highly error prone and it doesn't scale well.

### CODE REVIEW

File: com.unity.render-pipelines.core/Editor/Material/MaterialHeaderScopeList.cs [hunk 1]
```diff
@@ -27,9 +28,16 @@ public MaterialHeaderScopeList(uint defaultExpandedState = uint.MaxValue)
         /// <param name="title"><see cref="GUIContent"/> The title of the scope</param>
         /// <param name="expandable">The mask identifying the scope</param>
         /// <param name="action">The action that will be drawn if the scope is expanded</param>
-        public void RegisterHeaderScope(GUIContent title, uint expandable, Action<Material> action)
+        public void RegisterHeaderScope<TEnum>(GUIContent title, TEnum expandable, Action<Material> action)
```
Comment (svens-unity) [inline com.unity.render-pipelines.core/Editor/Material/MaterialHeaderScopeList.cs:31]: An enum might not necessarily be represented by an 4 byte variable. 
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/enum

File: com.unity.render-pipelines.core/Runtime/Documentation.cs [hunk 2]
```diff
@@ -79,4 +81,27 @@ class Documentation : DocumentationInfo
...
+    public static class DocumentationUtils
+    {
+        /// <summary>
+        /// Obtains the help url from an enum
+        /// </summary>
+        /// <typeparam name="TEnum">The enum with a <see cref="HelpURLAttribute"/></typeparam>
+        /// <param name="mask">The current value of the enum</param>
+        /// <returns>The full url</returns>
+        public static string GetHelpURL<TEnum>(TEnum mask)
+            where TEnum : struct, IConvertible
```
Comment (svens-unity) [inline com.unity.render-pipelines.core/Runtime/Documentation.cs:97]:
```suggestion
            where TEnum : Enum
```

**Review by svens-unity — APPROVED** [review]
Some minor comments to consider. Also this is making a few changes to public API, have we considered the impact on the user?

File: com.unity.render-pipelines.core/Editor/CoreEditorUtils.cs [hunk 3]
```diff
@@ -278,6 +278,8 @@ public static bool DrawHeaderFoldout(string title, bool state, bool isBoxed = fa
         /// <param name="isBoxed"> [optional] is the eader contained in a box style ? </param>
         /// <param name="hasMoreOptions"> [optional] Delegate used to draw the right state of the advanced button. If null, no button drawn. </param>
         /// <param name="toggleMoreOptions"> [optional] Callback call when advanced button clicked. Should be used to toggle its state. </param>
+        /// <param name="documentationURL">[optional] Documentation page</param>
+        /// <param name="contextAction">[optional] Context Action CallBack </param>
```
[the remark this answers is not in the document]
Comment (alex-vazquez) [reply com.unity.render-pipelines.core/Editor/CoreEditorUtils.cs:282]: Changed, thanks

File: com.unity.render-pipelines.core/Runtime/Documentation.cs [hunk 4]
```diff
@@ -79,4 +81,27 @@ class Documentation : DocumentationInfo
...
+    /// </summary>
+    public static class DocumentationUtils
+    {
+        /// <summary>
+        /// Obtains the help url from an enum
+        /// </summary>
+        /// <typeparam name="TEnum">The enum with a <see cref="HelpURLAttribute"/></typeparam>
+        /// <param name="mask">The current value of the enum</param>
+        /// <returns>The full url</returns>
+        public static string GetHelpURL<TEnum>(TEnum mask)
```
[the remark this answers is not in the document]
Comment (alex-vazquez) [reply com.unity.render-pipelines.core/Runtime/Documentation.cs:96]: changed

File: com.unity.render-pipelines.core/Editor/Material/MaterialHeaderScopeList.cs (hunk 1 above)
Comment (alex-vazquez) [reply com.unity.render-pipelines.core/Editor/Material/MaterialHeaderScopeList.cs:31]: thanks for the info, I didn't think about this

Comment (svens-unity) [reply com.unity.render-pipelines.core/Editor/Material/MaterialHeaderScopeList.cs:31]: I looked into an issue around enums earlier today. Seems like Unity serialization does not support 8 byte enums, I am not sure about 2 bytes but I think that would not cause any problems with this.

File: com.unity.render-pipelines.core/Editor/CoreEditorUtils.cs [hunk 5]
```diff
@@ -278,6 +278,8 @@ public static bool DrawHeaderFoldout(string title, bool state, bool isBoxed = fa
         /// <param name="isBoxed"> [optional] is the eader contained in a box style ? </param>
         /// <param name="hasMoreOptions"> [optional] Delegate used to draw the right state of the advanced button. If null, no button drawn. </param>
         /// <param name="toggleMoreOptions"> [optional] Callback call when advanced button clicked. Should be used to toggle its state. </param>
+        /// <param name="documentationURL">[optional] A url that will be open when the help button on the header is pressed</param>
```
Comment (reallyfancy) [inline com.unity.render-pipelines.core/Editor/CoreEditorUtils.cs:281]: The URL that the Unity Editor opens when the user presses the help button on the header.

File: com.unity.render-pipelines.core/Editor/CoreEditorUtils.cs [hunk 6]
```diff
@@ -278,6 +278,8 @@ public static bool DrawHeaderFoldout(string title, bool state, bool isBoxed = fa
...
+        /// <param name="contextAction">[optional] The callback executed when the burger menu on the header is pressed</param>
```
Comment (reallyfancy) [inline com.unity.render-pipelines.core/Editor/CoreEditorUtils.cs:282]: The callback that the Unity Editor executes when the user presses the burger menu on the header.

**Review by reallyfancy — CHANGES_REQUESTED** [review]
Just a couple of minor docs fixes.

**Review by TheoWong-pixel — APPROVED** [review]

**Review by reallyfancy — APPROVED** [review]

**Review by phi-lira — APPROVED** [review]

**Review by hdb-unity — APPROVED** [review]
No regressions found in manual testing.

I will ask to discuss the feedback from testing, in case some items should be addressed.

**Pull request merged**